### PR TITLE
EASY-1007/1003/1128 : datasetState in AMD / rights in DC / UTF8

### PR DESCRIPTION
**Pull request by jo-pol**
fixes EASY-1007 + EASY-1003 + EASY-1128

#### When applied it will

* [x] add history when `datasetState` in `AMD` is changed
* [x] expect the stream-id, xml-tag and old value in the CSV file. It used to expect the first two on the command line, the simple implementation will cause multiple rewrites of the same stream if multiple changes are specified for one stream.
* [x] have more validation of the input file and streams to change
* [x] reject CSV files that are not UTF-8 encoded
* [x] provide additional unit tests, even for UTF8

#### Where should the reviewer @DANS-KNAW/easy start?

The new introduction in the readme. The transformation is isolated in an object of its own with new unit tests, the InputRecord class is heavily refactored.

#### How should this be manually tested?

Either deploy on deasy and copy `src/test/resources/deasy-*data.csv` to `easy-dtap/shared`, then run with (both will make a few updates before running into an error)

    easy-update-metadata-dataset /vagrant/shared/deasy-data.csv
    easy-update-metadata-dataset /vagrant/shared/deasy-data.csv

or run from the project root with

    ./run.sh --fedora-url http://deasy.dans.knaw.nl:8080/fedora src/test/resources/deasy-input.csv
    ./run.sh --fedora-url http://deasy.dans.knaw.nl:8080/fedora src/test/resources/deasy-input.csv

It prompts for credentials which you can find on deasy with

    egrep 'fedora.admin' /opt/easy-webui/cfg/application.properties

#### related pull requests on github
repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

### CODE REVIEW

File: src/main/scala/nl/knaw/dans/easy/umd/Transformer.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+
+    new RuleTransformer(new RewriteRule {
+      override def transform(n: Node): Seq[Node] = n match {
+        case Elem(prefix, "datasetState", attribs, scope, _) =>
+          Elem(prefix, "datasetState", attribs, scope, false, Text(newState))
+        case Elem(prefix, "previousState", attribs, scope, _) =>
+          Elem(prefix, "previousState", attribs, scope, false, Text(oldState))
+        case Elem(prefix, "lastStateChange", attribs, scope, _) =>
+          Elem(prefix, "lastStateChange", attribs, scope, false, Text(DateTime.now().toString))
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/umd/Transformer.scala:70]: I suppose this toString causes one of the compiler warnings.

Comment (jo-pol) [reply src/main/scala/nl/knaw/dans/easy/umd/Transformer.scala:70]: Seem to need an explicit time zone to for platform independency and fix the travis build. Wonder how it is defined in the business layer of the web-ui.

File: src/main/scala/nl/knaw/dans/easy/umd/CommandLineOptions.scala [hunk 2]
```diff
@@ -19,19 +19,16 @@ import java.io.File
...
 import org.rogach.scallop._
-import org.slf4j.LoggerFactory
-
-import scala.util.{Failure, Success, Try}
+import org.slf4j.{Logger, LoggerFactory}
 
 class CommandLineOptions(args: Array[String] = "-ss -tt -f http:// src/test/resources/deasy-input.csv".split(" ")) extends ScallopConf(args) {
 
   appendDefaultToDescription = true
   editBuilder(_.setHelpWidth(110))
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/umd/CommandLineOptions.scala:28]: help width is 6 characters too wide for an iPad

File: src/main/scala/nl/knaw/dans/easy/umd/Command.scala [hunk 3]
```diff
@@ -38,36 +37,25 @@ object Command {
...
-  def run(implicit ps: Parameters): Try[Unit] = {
-    for {
-      records <- parse(ps.input)
-      _ <- records.map(update).find(_.isFailure).getOrElse(Success(())) // fail fast, if an error occurs, stop updating the rest of the stream!
-    } yield records.size
-  }
+  def run(implicit ps: Parameters): Try[Unit] = for {
+    records <- parse(ps.input)
+    _ <- records.map(update).find(_.isFailure).getOrElse(Success(())) // fail fast, if an error occurs, stop updating the rest of the stream!
+  } yield records.size
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/umd/Command.scala]: In test mode you might not (always) want to fail fast when the transformations detect problems with the content of a stream of a particular dataset.

**Comment by jo-pol** [comment]
Not yet tested from the command line but otherwise ready for review.

File: src/main/scala/nl/knaw/dans/easy/umd/Command.scala [hunk 4]
```diff
@@ -19,55 +19,46 @@ import com.yourmediashelf.fedora.client.FedoraClient
...
 object Command {
-  val log = LoggerFactory.getLogger(getClass)
+  val log: Logger = LoggerFactory.getLogger(getClass)
 
   def main(args: Array[String]): Unit = {
     log.debug("Starting command line interface")
     val ps = cmd.parse(args)
-    FedoraRequest.setDefaultClient(new FedoraClient(ps.fedoraCredentials))
     run(ps) match {
       case Success((count)) => log.info(s"Changed $count ${ps.streamID} streams.")
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/umd/Command.scala]: The count of updated records is not properly logged.

**Comment by jo-pol** [comment]
So far we designed the application to have the stream-id and xml-tag to be changed on the command line and the dataset id and old/new values in the csv. Perhaps it is better to have the stream-id and tag in the csv too. 

The change would allow to change EMD and DC in one batch. Now we are forced to work with different batches and when one of them halfway runs into trouble we have a whole bunch of datasets with EMD not matching DC. Within a single batch we can reduce the inconsistent state to a single dataset.

The easiest solution to implement would be a format like:

    FEDORA_ID,STREAM_ID,XML_TAG,NEW_VALUE,OLD_VALUE
    easy-dataset:998,EMD,accessRights,GROUP_ACCESS,OPEN_ACCESS
    easy-dataset:998,DC,rights,GROUP_ACCESS,OPEN_ACCESS
    easy-dataset:999,EMD,accessRights,GROUP_ACCESS,OPEN_ACCESS
    easy-dataset:999,DC,rights,GROUP_ACCESS,OPEN_ACCESS

A more user friendly format would be to allow to repeat the columns STREAM_ID,XML_TAG,NEW_VALUE,OLD_VALUE

The second format would allow (but doesn't necessarily needs to) to reduce the risk of an inconsistent state to a broken connection between updating the streams for one dataset. This potential additional safety leads to questions about error handling options, such as: should we only update if all streams/tags are actually changed or should we abort when just some of the streams/tags are changed.

**Comment by janvanmansum** [comment]
@jo-pol : please, run the build before you mark your PR as ready for merging.

File: src/main/scala/nl/knaw/dans/easy/umd/Transformer.scala (hunk 1 above)
Comment (janvanmansum) [reply src/main/scala/nl/knaw/dans/easy/umd/Transformer.scala:70]: @jo-pol:
The warnings are caused by a missing dependency. I have solved this before for `easy-stage-dataset` and other projects by just adding the dependency (even though it is apparently not used at runtime). See this fragment from `easy-staged-dataset`'s `pom.xml`:
```xml
        <dependency>
            <!-- Necessary to get rid of jodatime warnings about missing FromString
                (Jodatime is a transitive dependency through yourmediashelf)  -->
            <groupId>org.joda</groupId>
            <artifactId>joda-convert</artifactId>
        </dependency>
```

Please, apply this solution to your PR, so I can merge it.

**Review by janvanmansum — CHANGES_REQUESTED** [review]
See my comment about the build warnings.

**Review by janvanmansum — CHANGES_REQUESTED** [review]
I am getting the following error:
```
16:27:25.261 [main] ERROR nl.knaw.dans.easy.umd.Command$ - Failed Parameters(false,fedoraURL=http://localhost:8080/fedora, user=fedoraAdmin,/vagrant/shared/deasy-input.csv)
java.lang.Exception: failed to process: InputRecord(5,easy-dataset:4,EMD,datasetState,HISTORY,ANY_DISCIPLINE), reason: could not find EMD <HISTORY>HISTORY</HISTORY>```

**Comment by jo-pol** [comment]
I see only a MaxPermSize warning in the travis build and the other change request is addressed in https://github.com/DANS-KNAW/easy-update-metadata-dataset/pull/5/commits/8d7b5c82d58455972bf646ad77dc36a211d882ad

**Pull request merged**